### ISSUE TITLE
sdl2_ttf, opengl and SDL video driver

### ISSUE THREAD

**Issue by substring** [midwan/amiberry#262]
sdl2-ttf only works on OpenGL, which forces quite some changes on the OS as mentionned in the wiki ... This dependancy forces the whole system to run on KMS (at least on Pi). Otherwise, we're good to go with the pretty outdated SDL1 ...

I'd really love to have some full SDL2 compatibility not relying on OpenGL as well as setting the Pi video driver for the whole system to (f)KMS when it can have some bad consequences on other apps ...

One trick eventually (although not tested): you can force the SDL video driver with some environment variables : `SDL_VIDEODRIVER=KMSDRM`or `SDL_VIDEODRIVER=RPI` which could be one way not to force SDL2 to solely rely on KMS, thus untouching any side emulator. See:
- https://hg.libsdl.org/SDL/file/2088cd828335/src/video/SDL_video.c#l491 for the env var read and all the ifdef above to add the video drivers
- https://hg.libsdl.org/SDL/file/2088cd828335/src/video/kmsdrm/SDL_kmsdrmvideo.c#l162 for KMSDRM value
- https://hg.libsdl.org/SDL/file/2088cd828335/src/video/raspberry/SDL_rpivideo.c#l133 for the RPI part

I hope this can be of any help, close the issue at your will :)

**Comment by midwan** [comment]
@substring 
I've tried using only the `rpi` back-end in SDL2, but it seems to clash with the usage of Dispmanx that we have during emulation (for performance reasons, we open Dispmanx directly for the emulation screen, only the GUI is pure SDL2 in that target). It would seems that you cannot initialize and use Dispmanx after you've started an SDL2 application that uses the `rpi` back-end (which also uses Dispmanx).

That problem is not there with the `kmsdrm` back-end, so that is currently the only way to make this work on the Pi. And that requires the `fkms` driver in turn.

However, using the `fkms` driver I can also run the SDL1 apps normally, apps that use Dispmanx directly, etc. I haven't seen a problem with it (yet, at least).

What problems did you find with it?

**Comment by substring** [comment]
One background word before: i'm one of Recalbox devs. The reason i made this issue is that Recalbox solely displays through FBDEV, no X11, no KMS. So no OpenGL, OpenGLES at best. I won't be able to switch to SDL2 on pi when bumping to your last stable because of sdl2_ttf that does require OpenGL extensions.

Any other board we've ported Recalbox to is also using FBDEV (except x86). I haven't had time yet to check for odroid XU4 if it can compile sdl2_ttf.

And I can't change the whole Recalbox to use (f)kms, this would mean too much work for now, without being sure every emulator would support it ...

In the end, I'll have to stick to SDL1 :disappointed:

**Comment by midwan** [comment]
@substring 
I understand the problem, but I'm not sure I can do much about it as it is... :-/

The XU4 target does work (using SDL2), but since I don't that board here to test with, I do not know the details. We've had a few users reporting results only.

The SDL1 version is still there in the current version however, and will stick around at least as long as RetroPie is stuck on Jessie. But I'd like to get rid of it in the long run, to simplify the code internally and be able to use some of the features that are unique in SDL2 without having to re-implement them from scratch for SDL1 also... :)

**Comment by substring** [comment]
@midwan I think retropie most probably has the same problem as me regarding the upgrade. I can't tell for sure, but they may not have switched to KMS either

Regarding XU4: i guess it does work with SDL2 without using dispmanx ? So afterall, it's possible to use SDL2 without dispmanx ? I may as well check the makefile for compilation flags to see that. But anyway, even on XU4, Recalbox still uses FBDEV, not sure if the XU4 GPU driver has (as of now, i'm not talking about their GBM driver, but the good old FBDEV one) OpenGL extensions.

So let's close it for now ? Just wanted to stress out the consequences of using sdl2_ttf

**Comment by midwan** [comment]
@substring 
Yes, XU4 uses pure SDL2 (no Dispmanx), since Dispmanx is a RPI-only feature. :)
There are RPI targets that don't use Dispmanx as well, but the performance is worse there, that's why we recommend the other version instead.

Dispmanx is basically a (supposedly deprecated now) library that allows to control the VideoCoreIV GPU directly. That GPU is only on the RPI models of course, so other platforms cannot use it.

**Comment by substring** [comment]
SDL2 + --enable-video-rpi does use dispmanx, hence my very first comment ;)

**Comment by midwan** [comment]
@substring 
Yes, but that uses Dispmanx as a back-end from SDL2 itself, not from the emulator directly.

There are 2 different types of targets for the Raspberry Pi under SDL2:
- SDL2-Dispmanx: This uses SDL2 and the emulator also opens a Dispmanx screen during emulation directly, for maximum performance.
- SDL2: This uses whatever SDL2 provides and renders the emulation screen there. It depends on how you have configured SDL2 (so if you use --enable-video-rpi it will use Dispmanx). However, the performance on the RPI is significantly worse than the other approach.

I hope this helps in clearing it up :)
